rasa/nlu/data_augmentation/augmenter.py

```diff
@@ -0,0 +1,623 @@
+        classification_report: An existing classification report (without data augmentation).
+
+    Returns:
+        The set of intent names for which data augmentation will be performed.
+    """
+    # Determine low data, low performing and frequently confused intents
+    low_data_intents = sorted(
+        nlu_training_data.number_of_examples_per_intent.items(),
+        key=operator.itemgetter(1),
+    )[:num_intents_to_augment]
+    low_precision_intents = sorted(
+        [
+            (intent, classification_report[intent]["precision"])
+            for intent in nlu_training_data.intents
+        ],
+        key=operator.itemgetter(1),
+    )[:num_intents_to_augment]
+    low_recall_intents = sorted(
+        [
+            (intent, classification_report[intent]["recall"])
+            for intent in nlu_training_data.intents
+        ],
+        key=operator.itemgetter(1),
+    )[:num_intents_to_augment]
+    low_f1_intents = sorted(
+        [
+            (intent, classification_report[intent]["f1-score"])
+            for intent in nlu_training_data.intents
+        ],
+        key=operator.itemgetter(1),
+    )[:num_intents_to_augment]
+    freq_confused_intents = sorted(
+        [
+            (intent, sum(classification_report[intent]["confused_with"].values()))
+            for intent in nlu_training_data.intents
+        ],
+        key=operator.itemgetter(1),
+        reverse=True,
+    )[:num_intents_to_augment]
+
+    pooled_intents = {
+        intent_score_tuple[0]
+        for intent_score_tuple in low_data_intents
+        + low_precision_intents
+        + low_recall_intents
+        + low_f1_intents
+        + freq_confused_intents
+    }
+
+    return pooled_intents
+
+
+def create_paraphrase_pool(
+    paraphrases: TrainingData,
+    pooled_intents: Set[Text],
+    paraphrase_quality_threshold: float,
+) -> Dict[Text, List]:
+    """Determines all suitable paraphrases for data augmentation for the given intents.
+
+    Args:
+        paraphrases: The paraphrases for data augmentation.
+        pooled_intents: The intents for which to perform data augmentation.
+        paraphrase_quality_threshold: Accept/Reject threshold for individual paraphrases.
+
+    Returns:
+        The pool of suitable paraphrases for data augmentation.
+    """
+
+    paraphrase_pool = collections.defaultdict(list)
+    for p in paraphrases.intent_examples:
+        if p.data["intent"] not in pooled_intents or "metadata" not in p.data:
+            continue
+
+        paraphrases_for_example = p.data["metadata"]["example"]["paraphrases"]
+        paraphrase_scores = p.data["metadata"]["example"]["scores"]
+
+        for paraphrase, score in zip(paraphrases_for_example, paraphrase_scores):
+            if paraphrase == "" or float(score) < paraphrase_quality_threshold:
+                continue
+
+            paraphrase_pool[p.data["intent"]].append(
+                (p, set(paraphrase.lower().split()), paraphrase)
```

Comment:
I don't think we should create the vocabulary of paraphrases inside this function.